### PR TITLE
Support tunnel traffic QoS remapping

### PR DESCRIPTION
**Pull request by bingwang-ms**
Signed-off-by: bingwang <bingwang@microsoft.com>

**What I did**
This PR is to update `orchagent` to support tunnel traffic remapping.
HLD https://github.com/Azure/SONiC/pull/950
Changes include
- Change the termination type  of mux tunnel to P2P to apply different attributes
- Update `tunneldecaporch` to apply new SAI attributes to tunnel
- Update `muxorch` to apply new SAI attributes to tunnel
- Update `qosorch` to parse new QoS map tables

 
**Why I did it**
Please check HLD https://github.com/Azure/SONiC/pull/950 for more details.

**How I verified it**
Verified by vs test
```
sudo pytest test_mux.py -v --dvsname=vs   
collected 8 items                                                                                                                                                                                     

test_mux.py::TestMuxTunnel::test_Tunnel PASSED                                                                                                                                                  [ 12%]
test_mux.py::TestMuxTunnel::test_Peer PASSED                                                                                                                                                    [ 25%]
test_mux.py::TestMuxTunnel::test_Neighbor PASSED                                                                                                                                                [ 37%]
test_mux.py::TestMuxTunnel::test_Fdb PASSED                                                                                                                                                     [ 50%]
test_mux.py::TestMuxTunnel::test_Route PASSED                                                                                                                                                   [ 62%]
test_mux.py::TestMuxTunnel::test_acl PASSED                                                                                                                                                     [ 75%]
test_mux.py::TestMuxTunnel::test_mux_metrics PASSED                                                                                                                                             [ 87%]
test_mux.py::test_nonflaky_dummy PASSED                                                                                                                                                         [100%]
```
```
sudo pytest test_tunnel.py -v --dvsname=vs
collected 6 items                                                                                                                                                                                     

test_tunnel.py::TestDecapTunnel::test_TunnelDecap_v4 PASSED                                                                                                                                     [ 16%]
test_tunnel.py::TestDecapTunnel::test_TunnelDecap_v6 PASSED                                                                                                                                     [ 33%]
test_tunnel.py::TestDecapTunnel::test_TunnelDecap_MuxTunnel PASSED                                                                                                                              [ 50%]
test_tunnel.py::TestSymmetricTunnel::test_TunnelSymmetric_v4 PASSED                                                                                                                             [ 66%]
test_tunnel.py::TestSymmetricTunnel::test_TunnelSymmetric_v6 PASSED                                                                                                                             [ 83%]
test_tunnel.py::test_nonflaky_dummy PASSED  
```
System test is work in progress.

**Details if related**

### CODE REVIEW

File: orchagent/tunneldecaporch.cpp [hunk 1]
```diff
@@ -156,15 +169,47 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
...
+                else if (fvField(i) == DECAP_TC_TO_PG_MAP)
+                {
+                    string tc_to_pg_map_name = fvValue(i);
+                    // TODO: Validate TC_TO_PG_MAP name, and get map id
+                    if (exists)
+                    {
+                        setTunnelAttribute(fvField(i), tc_to_pg_map, tunnel_id);
+                    }
+                }
+                else if (fvField(i) == ENCAP_TC_TO_DSCP_MAP)
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:193]: Lets remove the ENCAP fields from this file as it is only handling decap tunnel here

File: orchagent/tunneldecaporch.cpp [hunk 2]
```diff
@@ -40,19 +40,36 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
         string key = kfvKey(t);
         string op = kfvOp(t);
 
-        IpAddresses ip_addresses;
-        IpAddress src_ip;
+        IpAddresses dst_ip_addresses;
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:43]: Please retain the same variable name.

File: orchagent/tunneldecaporch.cpp [hunk 3]
```diff
@@ -72,18 +89,14 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
...
-                        ip_addresses = IpAddresses(fvValue(i));
+                        dst_ip_addresses = IpAddresses(fvValue(i));
                     }
                     catch (const std::invalid_argument &e)
                     {
                         SWSS_LOG_ERROR("%s", e.what());
                         valid = false;
                         break;
                     }
-                    if (exists)
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:83]: Please keep the logic as before

File: orchagent/tunneldecaporch.cpp [hunk 4]
```diff
@@ -803,3 +920,25 @@ IpAddresses TunnelDecapOrch::getDstIpAddresses(std::string tunnelKey)
 
     return tunnelTable[tunnelKey].dst_ip_addrs;
 }
+
+sai_object_id_t TunnelDecapOrch::getTCToDSCPMap(std::string tunnelKey) const
+{
+    auto iter = tunnelTable.find(tunnelKey);
+    if (iter == tunnelTable.end())
+    {
+        SWSS_LOG_ERROR("Tunnel config not found %s", tunnelKey.c_str());
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:929]: Please have it as WARN. It is possible for tunnels not have the mapper and we don't want to raise ERR

File: orchagent/tunneldecaporch.cpp [hunk 5]
```diff
@@ -510,43 +624,45 @@ bool TunnelDecapOrch::setTunnelAttribute(string field, string value, sai_object_
...
 {
     TunnelEntry *tunnel_info = &tunnelTable.find(key)->second;
 
     // make a copy of tunnel_term_info to loop through
     vector<TunnelTermEntry> tunnel_term_info_copy(tunnel_info->tunnel_term_info);
 
     tunnel_info->tunnel_term_info.clear();
-    tunnel_info->dst_ip_addrs = new_ip_addresses;
-
+    tunnel_info->dst_ip_addrs = new_dst_ip_addresses;
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:643]: Please have a section in the beginning for p2p and retain the previous code as-is

File: orchagent/tunneldecaporch.cpp [hunk 6]
```diff
@@ -40,19 +40,36 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
         string key = kfvKey(t);
         string op = kfvOp(t);
 
-        IpAddresses ip_addresses;
-        IpAddress src_ip;
+        IpAddresses dst_ip_addresses;
+        IpAddress src_ip("0.0.0.0");
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:44]: Where are we getting the src_ip? If its from the DB data, then P2P can be determined based on the presence of src_ip and not the MUX tunnel name

**Review by prsunny — CHANGES_REQUESTED** [review]

**Comment by prsunny** [comment]
We would need VS test update and also reminder to change the dual tor sonic-mgmt test for IPinIP cases.

File: orchagent/tunneldecaporch.cpp (hunk 1 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:193]: The reason for I put it here is to parse entry from config db and store the map id into `TunnelTable`. The map will be used in `muxorch`. So I added a public inerface ` getTCToDSCPMap` and `getTCToQueueMap`.

File: orchagent/tunneldecaporch.cpp (hunk 6 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:44]: The src_ip is from config db. I was also thinking to determine the term type based on if there is `src_ip` or not. But I'm afraid we will also have `src_ip` for P2MP in the future. That's why I used the hardcoded mux tunnel name to make the decision.

File: orchagent/tunneldecaporch.cpp (hunk 3 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:83]: The logic has to be changed as we have to wait `src_ip` filed be parsed. That why I move `setIpAttribute` after thr for loop.

File: orchagent/tunneldecaporch.cpp (hunk 6 above)
Comment (prsunny) [reply orchagent/tunneldecaporch.cpp:44]: I think we can design to not have src_ip with P2MP.

Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:44]: Sure. I will update the logic here if that is the case. Thanks

**Comment by bingwang-ms** [comment]
PR still WIP. Thanks

File: orchagent/qosorch.cpp [hunk 7]
```diff
@@ -103,6 +106,14 @@ task_process_status QosMapHandler::processWorkItem(Consumer& consumer)
     SWSS_LOG_ENTER();
 
     sai_object_id_t sai_object = SAI_NULL_OBJECT_ID;
+    /*
+     ToDo:
+     As we are going to have more than one QosMap in type DSCP_TO_TC_MAP, TC_TO_PRIORITY_GROUP_MAP and TC_TO_QUEUE_MAP,
```
Comment (neethajohn) [inline orchagent/qosorch.cpp:111]: As discussed offline, we do not need to apply any maps at the switch level. Please update the PR to create separate qos maps for the tables that contain the tunnel mappings

**Comment by bingwang-ms** [comment]
> PR still WIP. Thanks

PR is ready for review.

File: orchagent/muxorch.cpp [hunk 8]
```diff
@@ -1154,7 +1188,9 @@ MuxOrch::MuxOrch(DBConnector *db, const std::vector<std::string> &tables,
          Orch2(db, tables, request_),
          decap_orch_(decapOrch),
          neigh_orch_(neighOrch),
-         fdb_orch_(fdbOrch)
+         fdb_orch_(fdbOrch),
+         cfgTunnelTable_(db, CFG_TUNNEL_TABLE_NAME)
```
Comment (prsunny) [inline orchagent/muxorch.cpp:1192]: Why do we use CFG table here? All the attributes are available from APP_DB table, right? Also, please use the naming convention as rest of the params. i.e use `_ `seperator.

File: orchagent/tunneldecaporch.cpp (hunk 2 above)
Comment (prsunny) [reply orchagent/tunneldecaporch.cpp:44]: Please address this comment

File: orchagent/tunneldecaporch.cpp [hunk 9]
```diff
@@ -40,19 +41,29 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
         string key = kfvKey(t);
         string op = kfvOp(t);
 
-        IpAddresses ip_addresses;
-        IpAddress src_ip;
+        IpAddresses dst_ip_addresses;
+        IpAddress src_ip_address("0.0.0.0");
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:45]: Why is this changed. Lets keep the original logic

File: orchagent/tunneldecaporch.cpp [hunk 10]
```diff
@@ -5,6 +5,7 @@
 #include "crmorch.h"
 #include "logger.h"
 #include "swssnet.h"
+#include "qosorch.h"
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:8]: Please remove this header file

File: orchagent/tunneldecaporch.cpp [hunk 11]
```diff
@@ -40,19 +41,29 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
...
+        IpAddress src_ip_address("0.0.0.0");
         IpAddress* p_src_ip = nullptr;
         string tunnel_type;
         string dscp_mode;
         string ecn_mode;
         string encap_ecn_mode;
         string ttl_mode;
+        sai_object_id_t dscp_to_dc_map_id = SAI_NULL_OBJECT_ID;
+        sai_object_id_t tc_to_pg_map_id = SAI_NULL_OBJECT_ID;
+        TunnelTermType term_type = TUNNEL_TERM_TYPE_P2MP;
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:54]: I don't think we need to have this variable assigned here. In the calling function, based on p_src_ip, this can be determined. Please avoid and extra variable being passed

File: orchagent/muxorch.cpp [hunk 12]
```diff
@@ -1208,6 +1244,47 @@ bool MuxOrch::handleMuxCfg(const Request& request)
...
+// resolve the ids from QosOrch
+bool MuxOrch::resolveQosTableIds()
+{
+    std::vector<FieldValueTuple> field_value_tuples;
+    if (cfgTunnelTable_.get(MUX_TUNNEL, field_value_tuples))
+    {
+        KeyOpFieldsValuesTuple tuple{"TUNNEL", MUX_TUNNEL, field_value_tuples};
+        for (auto it = kfvFieldsValues(tuple).begin(); it != kfvFieldsValues(tuple).end(); it++)
+        {
+            if (qos_to_ref_table_map.find(fvField(*it)) != qos_to_ref_table_map.end())
```
Comment (prsunny) [inline orchagent/muxorch.cpp:1257]: We cannot use another orch's Data structure here. This must be handled in qosorch and retrieve from there.

File: orchagent/tunneldecaporch.cpp [hunk 13]
```diff
@@ -327,8 +389,8 @@ bool TunnelDecapOrch::addDecapTunnel(string key, string type, IpAddresses dst_ip
 
     tunnelTable[key] = { tunnel_id, overlayIfId, dst_ip, {} };
 
-    // create a decap tunnel entry for every ip
-    if (!addDecapTunnelTermEntries(key, dst_ip, tunnel_id))
+    // create a decap tunnel entry for every source_ip - dest_ip pair
+    if (!addDecapTunnelTermEntries(key, src_ip, dst_ip, tunnel_id, term_type))
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:393]: You can pass it as `*p_src_ip` and retain the existing variables.

File: orchagent/tunneldecaporch.cpp [hunk 14]
```diff
@@ -372,19 +443,38 @@ bool TunnelDecapOrch::addDecapTunnelTermEntries(string tunnelKey, IpAddresses ds
...
+            key = dst_ip;
+        }
 
-        // check if the there's an entry already for the ip
-        if (existingIps.find(ip) != existingIps.end())
+        // check if the there's an entry already for the key pair
+        if (existingIps.find(key) != existingIps.end())
         {
-            SWSS_LOG_NOTICE("%s already exists. Did not create entry.", ip.c_str());
+            SWSS_LOG_ERROR("%s already exists. Did not create entry.", key.c_str());
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:477]: Please revert back to NOTICE. This was part of a repair item

File: orchagent/tunneldecaporch.cpp [hunk 15]
```diff
@@ -803,3 +940,35 @@ IpAddresses TunnelDecapOrch::getDstIpAddresses(std::string tunnelKey)
...
+ *
+ * Arguments:
+ *    @param[in] tunnle_name - The name of tunnel
+ *    @param[in] map_type_name - The type of referenced QoS map
+ *    @param[in] tuple - The KeyOpFieldsValuesTuple that contains keys - values
+ *
+ * Return Values:
+ *    @return The sai_object_id of referenced map, or SAI_NULL_OBJECT_ID  if there's an error
+ */
+sai_object_id_t TunnelDecapOrch::resolveQosMapId(std::string tunnle_name, std::string map_type_name, KeyOpFieldsValuesTuple& tuple)
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:956]: We should have qos related functions in qosorch. Please consider this for qosorch

File: orchagent/tunneldecaporch.cpp [hunk 16]
```diff
@@ -803,3 +940,35 @@ IpAddresses TunnelDecapOrch::getDstIpAddresses(std::string tunnelKey)
...
+{
+    sai_object_id_t id;
+    string object_name;
+    ref_resolve_status status = resolveFieldRefValue(QosOrch::getTypeMap(), map_type_name, qos_to_ref_table_map.at(map_type_name), tuple, id, object_name);
```
Comment (prsunny) [inline orchagent/tunneldecaporch.cpp:960]: As mentioned, it is not a good idea to use qos data structure in another orch

**Review by prsunny — CHANGES_REQUESTED** [review]

File: orchagent/muxorch.cpp (hunk 8 above)
Comment (bingwang-ms) [reply orchagent/muxorch.cpp:1192]: We need to read `encap_tc_to_queue_map` and `encap_tc_color_to_dscp_map` from `TUNNEL` table.
```
"TUNNEL": {
        "MuxTunnel0": {
            "encap_tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE_TUNNEL]",
            "encap_tc_color_to_dscp_map": "[TC_TO_DSCP_MAP|AZURE_TUNNEL]"
        }
    }
```
Yes, these attributes are also in `APP_DB`. But how can we ensure that the `TUNNEL` table is already ported into `APP_DB` when creating muxtunnel? It's `tunnelmgrd` who write the `TUNNEL` table into `APP_DB`, right?

The name of variable will be corrected. Thanks

File: orchagent/tunneldecaporch.cpp (hunk 13 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:393]: Yes, we can do that. But we have to check if p_src_ip is NULL. So I think adding another variable is better here.

File: orchagent/tunneldecaporch.cpp (hunk 10 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp:8]: The `qosorch.h` header is needed as we need to access `gQosOrch`

File: orchagent/tunneldecaporch.cpp (hunk 9 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp]: Fixed. Thanks.

File: orchagent/tunneldecaporch.cpp (hunk 11 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp]: Fixed. The variable `term_type` is removed now. But `dscp_to_dc_map_id ` and `tc_to_pg_map_id` are required and can't be removed.

File: orchagent/tunneldecaporch.cpp (hunk 15 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp]: Fixed.

File: orchagent/tunneldecaporch.cpp (hunk 16 above)
Comment (bingwang-ms) [reply orchagent/tunneldecaporch.cpp]: Fixed.

File: orchagent/muxorch.cpp [hunk 17]
```diff
@@ -1183,13 +1184,13 @@ void MuxOrch::update(SubjectType type, void *cntx)
...
-MuxOrch::MuxOrch(DBConnector *db, const std::vector<std::string> &tables,
+MuxOrch::MuxOrch(DBConnector *cfg_db, DBConnector *app_db, const std::vector<std::string> &tables,
          TunnelDecapOrch* decapOrch, NeighOrch* neighOrch, FdbOrch* fdbOrch) :
-         Orch2(db, tables, request_),
+         Orch2(cfg_db, tables, request_),
          decap_orch_(decapOrch),
          neigh_orch_(neighOrch),
          fdb_orch_(fdbOrch),
-         cfgTunnelTable_(db, CFG_TUNNEL_TABLE_NAME)
+         app_decap_tunnel_table_(app_db, APP_TUNNEL_DECAP_TABLE_NAME)
```
Comment (prsunny) [inline orchagent/muxorch.cpp:1193]: I think there is a gap here. Let us discuss offline

File: orchagent/tunneldecaporch.cpp (hunk 13 above)
Comment (prsunny) [reply orchagent/tunneldecaporch.cpp:391]: ok

File: orchagent/muxorch.cpp (hunk 12 above)
Comment (prsunny) [reply orchagent/muxorch.cpp]: Agree

File: orchagent/muxorch.cpp (hunk 17 above)
Comment (bingwang-ms) [reply orchagent/muxorch.cpp]: Fixed as discussion. Please help take a look. Thanks

File: orchagent/muxorch.cpp [hunk 18]
```diff
@@ -1229,10 +1264,33 @@ bool MuxOrch::handlePeerSwitch(const Request& request)
...
         auto it =  dst_ips.getIpAddresses().begin();
         const IpAddress& dst_ip = *it;
-        mux_tunnel_id_ = create_tunnel(&peer_ip, &dst_ip);
+
+        // Read dscp_mode of MuxTunnel0 from decap_orch
+        string dscp_mode_name = decap_orch_->getDscpMode(MUX_TUNNEL);
+        if (dscp_mode_name == "")
+        {
+            SWSS_LOG_INFO("dscp_mode for tunnel %s is not available yet", MUX_TUNNEL);
+            return false;
```
Comment (prsunny) [inline orchagent/muxorch.cpp:1275]: We do not have to check for these new attributes. The dst_ips get will ensure MUX_TUNNEL is created. We can have these attributes as optional. We don't have to worry about test failures if the MUX_TUNNEL is not populated right away.

**Review by prsunny — COMMENTED** [review]
Overally, the changes lgtm. One minor comment regarding backward compatibility .

File: orchagent/muxorch.cpp (hunk 18 above)
Comment (bingwang-ms) [reply orchagent/muxorch.cpp]: Fixed. Thanks

**Review by prsunny — APPROVED** [review]
